lib/controllers/ephemeral/pin_entry_controller.dart

```diff
@@ -0,0 +1,82 @@
+import 'dart:async';
+import 'package:flutter/cupertino.dart';
+import 'package:get/get.dart';
+import 'package:flutter_secure_storage/flutter_secure_storage.dart';
+import 'package:pin_code_fields/pin_code_fields.dart';
+import 'package:pispapp/controllers/ephemeral/local_auth_controller.dart';
+import 'package:pispapp/routes/app_pages.dart';
+
+class PINEntryController extends GetxController {
+  // Constructor to fetch PIN automatically
+  PINEntryController() {
+    _fetchPIN();
+  }
+
+  // Constants
+  static const String key = 'USER_PIN';
+  static const Duration timeoutDuration = Duration(milliseconds: 1000);
+  static const int PINlength = 6;
+
+  // Storage
+  final FlutterSecureStorage _storage = const FlutterSecureStorage();
+
+  // Fields to be populated
+  String _correctPIN = '';
+  bool _userSetPIN = false;
+
+  // Returns whether or not the user set a PIN
+  bool get userSetPIN => _userSetPIN;
+
+  void onFallbackVerificationNeeded() {
+    Get.toNamed<dynamic>(Routes.PIN_ENTRY);
+  }
+
+  // Called after correct PIN has been populated
+  bool authenticate(String pin) => pin == _correctPIN;
+
+  // Authentication process
+  void onPINEntered(String pin, TextEditingController tc, StreamController<ErrorAnimationType> ec) {
+    if(!_userSetPIN) { return; } // Do nothing if there is no previous set PIN
+    tc.clear();
+    if (authenticate(pin)) {
+      // PIN entry screen always opened on top of something AND
+      // is not poppable via device controls so we can be sure it
+      // is still on top when we call this
+      Get.back();
+      Get.find<LocalAuthController>().verificationInProgress = false;
+    }
+    else {
+      // Shakes the entry fields to indicate incorrect PIN
+      ec.add(ErrorAnimationType.shake);
+    }
+  }
+
+  Future<void> _fetchPIN() async {
+    final String pin = await Future.any([
+      _storage.read(key: PINEntryController.key),
+      Future.delayed(PINEntryController.timeoutDuration, () => null)
+    ]);
+
+    if (pin == null) {
+      print('no previous PIN set');
+      _userSetPIN = false;
+    }
+    else {
+      print('fetched pin as $pin');
+      _userSetPIN = true;
+      _correctPIN = pin;
+    }
+    update();
+  }
+
+  Future<void> storeNewPIN(String pin) async {
+    print('storing new pin $pin');
+    await Future.any([
+      _storage.write(key: PINEntryController.key, value: pin),
+      Future.delayed(PINEntryController.timeoutDuration, () => null)
+    ]);
+
+    // Update new PIN
+    await _fetchPIN();
```

Comment:
What if it fails to store the pin? Throw an error perhaps?
Also, I think you don't need to fetch again here since you know the new value if it succeed?

Besides, if we want to fetch in case the user change the pin from somewhere else, I think it would be better to fetch it inside the `onPINEntered` function

Reply:
The write func already throws an error if it fails, but I added one in case a timeout occurs. 

The reason why I fetch afterwards is to ensure that what is stored is consistent with the client. The cost I feel is negligible and I think it helps to verify that the PIN was stored properly. What do you think?

Reply:
I think it would be better to also throw an error in case a timeout occurs instead of just doing nothing...
We need to notify the user that the PIN couldn't be saved for both cases